Add ALTS auth guide

jiangtaoli2016 opened the pull request.

jiangtaoli2016 commented:
@srini100 Could you please review?

chalin commented:
I'm thinking that it might make more sense to put the ALTS page under auth, as in `/docs/guides/auth/ALTS.md` rather than `/docs/guides/alts_auth.md`. Thoughts @jtattermusch @ejona86?

chalin commented:
Thanks for moving the new file.

Rather than create a new (and almost empty) index file (`/docs/guides/auth/_index.md`), you should rename `/docs/guides/auth.md` to `/docs/guides/auth/_index.md`.

jiangtaoli2016 commented:
Thanks @chalin for suggestion. PR Revised.

@srini100 @ejona86 @jtattermusch Can anyone of you review please?

chalin commented:
@mjpitz will be taking a look

chalin commented on content/docs/guides/auth/_index.md:
```diff
@@ -24,6 +24,12 @@ The following authentication mechanisms are built-in to gRPC:
+  [ALTS](https://cloud.google.com/security/encryption-in-transit/application-layer-transport-security)
+  as a transport security mechanism, if the application is running on
+  [Google Cloud Platform (GCP)](https://cloud.google.com/). See
+  [ALTS Authentication Guide](https://grpc.io/docs/guides/auth/ALTS) for
```
The link to the ALTS page should be just `[...](ALTS)`.

chalin replied:
Oups, my bad: the file name (`ALTS.md`) and the link (`ALTS`) should both be in lower case.

srini100 commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,246 @@
+or via
+[GCP console](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#using).
+
+Google issues an ALTS credential for each service account running on the GCE VM.
```
Instead of "Google issues an ALTS..", should this be "Google Cloud Platform issues an ALTS.."

srini100 commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,246 @@
+Google issues an ALTS credential for each service account running on the GCE VM.
+The ALTS credentials are securely located in the hypervisor. The private key of
+an ALTS credential is not accessible to the VM and the application. The session
+keys used for end-to-end encryption are exposed to the gRPC stack. Google fully
```
Same here. Google => Google Cloud Platform.

chalin commented:
When you have a chance, could you also rebase this PR -- it is at least 8 commits behind? Thanks.

srini100 commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,246 @@
+
+On a successful ALTS connection, the peer information (e.g., client’s service
+account) is stored in the AltsContext. gRPC provides a utility library for
+client authorization check. Assume that the server knows the expected client
```
Assume => Assuming

chalin commented:
Btw, this change makes the "Authentication" entry disappear from the sidenav, but I'll fix that in a followup PR. I'll also submit some copyedits at that time.

jiangtaoli2016 commented:
Thanks you @srini100 and @chalin for reviews and suggestions! I addressed all comments except 
"makes the Authentication entry disappear from the sidenav". @chalin could you please fix that once this PR gets merged?

chalin commented:
> @chalin could you please fix that once this PR gets merged?

Yes

mjpitz commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,246 @@
+---
```
So I think this should probably be broken down into two pages.
1. A general purpose guide describing how to use the ALTS feature in gRPC (platform agnostic)
2. A new blog post describing how to use the ALTS feature inside of Google Cloud Platform

Given that ALTS has an associated white-papper, I expect implementations to be cropping up any time. In a cursory search, I was able to find a relevant thread someone started on [Istio](https://discuss.istio.io/t/what-does-the-alts-handshaker-do/6072). Since we can expect other implementations to crop up, it would be helpful to decouple Google Cloud Platform specifics from the core feature guide.

@thisisnotapril and I were talking earlier about good content for blog posts. I think provider-specific examples are great candidates for this.

Curious what other peoples thoughts are on this.

mjpitz replied:
While the Istio example is referencing the client-side, I could see something like Hashicorp Vault supporting the server-side of ALTS.

jiangtaoli2016 replied:
Unfortunately, gRPC ALTS only works for GCP. It does not work on-prem or on other Clouds.

mjpitz replied:
Sure. What I'm suggesting is that it's only a matter of time until we see other implementations. Between the white paper and the client-side implementation, there's enough context to piece together alternative backend implementations.

jiangtaoli2016 replied:
Thanks @mjpitz for suggestion. I split the guide into two pages: one generic guide, the other is Google Cloud Platform page. I also made a note that ALTS is currently available on GCP. 

Technically, if someone writes an ALTS handshaker service, it can run outside GCP. It is out of scope of this guide.

mjpitz replied:
Oh I completely agree that is outside the scope of this change. It's just something I thought we should consider as we developed content around it.

jiangtaoli2016 replied:
Do you have specific suggestion on developing content? For ALTS on GCP or ALTS outside GCP?

mjpitz replied:
So the big thing that I try to consider is separating documentation of a given feature from implementation or use. In this case, ALTS (feature) and GCP/GKE (use case). This is [generally?] a good practice to follow when there are pluggable interfaces or configuration that allows backend implementation details to be changed out.

Let us consider the previous version where everything was in a single document. If someone like Vault came along and implemented an ALTS handshake service, its reasonable to expect an end-user to request documentation around this integration. We would then need to update to the ALTS documentation to detail both GCP's and Vaults support or remove the provider specifics entirely.

As the content currently stands, we get everything we wanted from the original (documentation of ALTS feature and it's integration with GCP/GKE) while being able to decouple provider-specific details from the core guide. I feel like the blog posts are a great way to showcase integrations with different providers, while keeping the core documentation agnostic.

jiangtaoli2016 replied:
Thanks for suggestion!

As for blog post, I got my leadership approval for ALTS user guide, but not blog post yet. I hope we can publish user guide first, so that user can start using this feature.

jiangtaoli2016 replied:
Let me remove GCP specific information from this guide. I agree it is not suitable in this guide.

mjpitz commented on content/docs/guides/auth/ALTS_GCP.md:
```diff
@@ -0,0 +1,30 @@
+---
```
I think this content would be more powerful as a blog post with more of a walkthrough.

mjpitz commented on content/docs/guides/auth/ALTS_GCP.md:
```diff
@@ -0,0 +1,30 @@
+changed using
+[gCloud command](https://cloud.google.com/sdk/gcloud/reference/compute/instances/set-service-account)
+or via
+[GCP console](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#using).
```
This is all very GCE centric. I assume the identity in Kubernetes is a given processes service account? It would be useful to clarify some of this in this guide.

jiangtaoli2016 replied:
I added additional info about GKE. ALTS works for GKE, only using underlying node identity (GKE node is GCE VM). Per-pod identity is WIP.

chalin commented:
> "makes the Authentication entry disappear from the sidenav". @chalin could you please fix that once this PR gets merged?

If you rebase, the sidenav should work now.

jiangtaoli2016 commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,231 @@
+### Client Authorization
+
+On a successful ALTS connection, the peer information (e.g., client’s service
+account) is stored in the AltsContext. gRPC provides a utility library for
```
I provide a link to AltsContext. AltsContext contains context about ALTS connection (e.g., local identity and peer identity, security level, and so on). It is available to both client and server side. For authorization, we just check peer identity against certain ACLs.

jiangtaoli2016 commented:
@mjpitz I un-split to guides into one ALTS guide and addressed all your comments. PTAL.

Once ALTS can be supported outside GCP, we are happy to split the guide into a generic ALTS guide and platform specific blog posts.

jiangtaoli2016 commented:
@mjpitz Can we move forward with this guide or you still have concerns?

chalin commented:
We might want to first re-discuss this PR during our Docs meeting next Monday.
@thisisnotapril @mjpitz - WDYT?

cesarghali commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,249 @@
+  "google.golang.org/grpc/credentials/alts"
+)
+
+altsTC := alts.NewClientCreds(alts.DefaultClientOptions())
```
This seems to be the client code :)

A server example can be found here:
https://github.com/grpc/grpc-go/blob/master/examples/features/encryption/ALTS/server/main.go

jiangtaoli2016 replied:
Good catch. Copy&paste mistake :(

cesarghali commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,249 @@
+### Server Authorization
+
+gRPC has built-in server authorization support using ALTS. A gRPC client using
+ALTS can set the expected service accounts prior to establishing a connection.
```
s/expected service accounts/expected server service accounts

cesarghali commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,249 @@
+
+gRPC has built-in server authorization support using ALTS. A gRPC client using
+ALTS can set the expected service accounts prior to establishing a connection.
+Then, at the end of the handshake, it is guaranteed that the server identity
```
Just a minor rewording:

```
Then, at the end of the handshake, server authorization guarantees
that the server identity matches one of the service accounts specified
by the client. Otherwise, the connection fails.
```

jiangtaoli2016 commented:
Thanks @cesarghali for reviews. Your comments are resolved.

jiangtaoli2016 commented:
@chalin @mjpitz Any outcome from the doc meeting?

chalin commented:
> Any outcome from the doc meeting?

Yes, @ejona86 will be contacting you shortly.

jiangtaoli2016 commented:
Per off-discussion communication with @ejona86 , I revised the guide to be platform generic. Indeed, the gRPC ALTS code is open-sourced and platform generic. 
@chalin @mjpitz  Could you please take a look and let me know what do you think?

In a followup, I will write GCP specific ALTS information somewhere (e.g., developers.google.com), and link from this ALTS guide.

chalin commented:
@jiangtaoli2016 - sounds good.
@mjpitz - do you think that you'll be able to give this another look before the week is out?

mjpitz commented:
I should be able to get to this either later today or tomorrow.

mjpitz commented on content/docs/guides/auth/ALTS.md:
```diff
@@ -0,0 +1,232 @@
+Note that ALTS is fully functional if the application runs on
+[Google Cloud Platform](https://cloud.google.com/). ALTS could be run on any
+platforms with a pluggable
+[ALTS handshaker service](https://github.com/grpc/grpc/blob/master/src/proto/grpc/gcp/handshaker.proto#L233-L243).
```
This should probably reference a specific commit sha instead of just the copy on the master branch.

Thoughts?

jiangtaoli2016 replied:
Good suggestion! Let me update in a separate PR.

chalin commented:
I'll merge as is and make a copy-editing pass after.
Thanks for the contribution @jiangtaoli2016, and for the reviews @mjpitz!

jiangtaoli2016 commented:
Thank you @chalin @mjpitz for reviewing!
